Opentable specific mods

travees opened the pull request:
Re-add OpenTable mods after upstream sync

travees committed:
More OpenTable mods

- Allow for exhibitor use
- Use archive module for tarball installs
- Update specs
- Add parameters to README

travees committed:
Don't manage if $exhibitor_managed is 'true'

SegFaultAX commented on manifests/config.pp:
```diff
@@ -106,7 +106,8 @@
   }
 
-  if $exhibitor_manaaged == 'false' {
+  if $exhibitor_managed == false {
```
Consider using `stdlib/validate_bool`

travees replied:
This is validated in init.pp.
Preferable to move it into config.pp?

SegFaultAX commented on manifests/install.pp:
```diff
@@ -68,33 +74,69 @@
       }
     }
   } else {
+    include '::archive'
```
Perhaps this should be

``` puppet
} else if ($install_method == 'archive') {
  # ...
} else {
  fail("you must select a valid installation method for zookeeper")
}
```

SegFaultAX commented on manifests/install.pp:
```diff
@@ -68,33 +74,69 @@
+
+    $basefilename = "zookeeper-${ensure}.tar.gz"
+    $package_url = "${mirror_url}/zookeeper/zookeeper-${ensure}/${basefilename}"
+    $checksum_url = "http://www-us.apache.org/dist/zookeeper/zookeeper-${ensure}/${basefilename}.sha1"
```
Perhaps the checksum should be passed in as an argument to this module. Otherwise an attacker poisoning the tarball merely has to also poison the checksum file and you'd never know if you downloaded a bad version.

SegFaultAX commented on manifests/install.pp:
```diff
@@ -68,33 +74,69 @@
+      ],
+    }
+
+    archive { "${package_dir}/${basefilename}":
```
Should this resource `notify` the service resource? (Or else, should the service resource subscribe to this resource). That way if the version is updated and zk is reinstalled, zk will be _optionally_ restarted.

travees replied:
Maybe? The deb doesn't do a restart on upgrade.

SegFaultAX commented on manifests/znode.pp:
```diff
@@ -2,16 +2,18 @@
 ) {
-  
+
+  $znode_check = "/bin/bash -c \"if [[ \\\"$(echo 'stat /${name}' | /opt/zookeeper/bin/zkCli.sh 2>&1 1>/dev/null)\\\" == 'Node does not exist'* ]]; then exit 1; else exit 0; fi\""
```
This hurts my soul. (No action required)

fessyfoo replied:
1.) rely on bash's `[[]]`'s exit code, remove extraneous inversion of exit using `if` and instead use `!=`,  avoid multi backslashes. 

```
$znode_check = "/bin/bash -c '[[ $(echo \"stat /${name}\" | /opt/zookeeper/bin/zkCli.sh 2>&1 1>/dev/null) != \"Node does not exist\"* ]]'"
```

2.) use posix shell's case to get wildcard match

```
$znode_check = "case $(echo 'stat /${name}' | /opt/zookeeper/bin/zkCli.sh 2>&1 1>/dev/null) in 'Node does not exist'*) true;; *) false ;; esac"
```

3.) use grep, rely on it's  exit code

```
$znode_check = "echo 'stat /${name}' | /opt/zookeeper/bin/zkCli.sh 2>&1 1>/dev/null | grep -vq 'Node does not exist'"
```

4.) write a command line utility, install it on the box with a file resource and call it from the unless hook. 

I prefer number 4,  if you feel like messing with it.

SegFaultAX commented on spec/classes/install_spec.rb:
```diff
@@ -275,4 +283,53 @@
+      :ensure => 'link',
+      :target => extract_path,
+    })}
+    it { should contain_archive("#{package_dir}/#{basefilename}").with({
```
Grr, not a fan of multiline blocks using `{`. Prefer `it do` for multiline blocks.

travees committed:
Feedback fixes #1

- Require $install_method to be either 'package' or 'archive'
- Pass in optional archive checksum instead of getting from external URL when using
  archive install method

travees committed:
Fix spec update README

spec didn't like unquoted hash keys

travees committed:
Make znode more useful

- Move exec conditional script into its own file
- Do some validation of znode path
- shellquote for safety
